src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionInvoker.cs

```diff
@@ -303,47 +321,99 @@ private Task Next(ref State next, ref Scope scope, ref object state, ref bool is
+            // vs just a Page. We need to know up front because we want to execute handler methods
+            // on the PageModel without instantiating the Page or ViewContext.
+            var hasPageModel = _actionDescriptor.HandlerTypeInfo != _actionDescriptor.PageTypeInfo;
+            if (hasPageModel)
             {
-                viewStarts = Array.Empty<IRazorPage>();
+                return ExecutePageWithPageModelAsync();
             }
             else
             {
-                viewStarts = new IRazorPage[CacheEntry.ViewStartFactories.Count];
-                for (var i = 0; i < viewStarts.Length; i++)
+                return ExecutePageWithoutPageModelAsync();
+            }
+        }
+
+        private async Task ExecutePageWithPageModelAsync()
+        {
+            // Since this is a PageModel, we need to activate it, and then run a handler method on the model.
+            //
+            // We also know that the model is the pagemodel at this point.
+            Debug.Assert(_actionDescriptor.ModelTypeInfo == _actionDescriptor.HandlerTypeInfo);
+            _model = CacheEntry.ModelFactory(_pageContext);
+            _pageContext.ViewData.Model = _model;
+
+            if (CacheEntry.PropertyBinder != null)
+            {
+                await CacheEntry.PropertyBinder(_pageContext, _model);
+            }
+
+            // This is a workaround for not yet having proper filter for Pages.
+            PageSaveTempDataPropertyFilter propertyFilter = null;
+            for (var i = 0; i < _filters.Length; i++)
+            {
+                propertyFilter = _filters[i] as PageSaveTempDataPropertyFilter;
+                if (propertyFilter != null)
                 {
-                    var pageFactory = CacheEntry.ViewStartFactories[i];
-                    viewStarts[i] = pageFactory();
+                    break;
                 }
             }
-            _pageContext.ViewStarts = viewStarts;
 
-            if (actionDescriptor.ModelTypeInfo == actionDescriptor.PageTypeInfo)
+            if (propertyFilter != null)
             {
-                _model = _page;
+                propertyFilter.Subject = _model;
+                propertyFilter.ApplyTempDataChanges(_pageContext.HttpContext);
             }
-            else
+
+            _result = await ExecuteHandlerMethod(_model);
+            if (_result is PageResult pageResult)
             {
-                _model = CacheEntry.ModelFactory(_pageContext);
+                // If we get here, we are going to render the page, so we need to create it and then initialize
+                // the context so we can run the result.
+                _viewContext = new ViewContext(
+                    _pageContext,
+                    NullView.Instance,
+                    _pageContext.ViewData,
+                    _tempDataFactory.GetTempData(_pageContext.HttpContext),
+                    TextWriter.Null,
```

Comment:
Should we make a ctor without the `TextWriter`? I don't think we initialize it to the right value either here or in `VIewExecutor`